File: libs/backend-apisix-standalone/src/typing.ts
import * as ADCSDK from '@api7/adc-sdk';
import { z } from 'zod';

export const ADC_UPSTREAM_SERVICE_ID_LABEL = '__ADC_UPSTREAM_SERVICE_ID';

const Metadata = {
  id: z.string(),
  name: z.string(),
  desc: z.string().optional(),
  labels: z.record(z.string(), z.string()).optional(),
};
const ModifiedIndex = {
  modifiedIndex: z.int(),
};
const Timeout = z.strictObject({
  connect: z.number().gt(0),
  send: z.number().gt(0),
  read: z.number().gt(0),
});
const Plugins = z.record(z.string(), z.record(z.string(), z.unknown()));
const Status = z.union([z.literal(0), z.literal(1)]);
const Port = z.int().min(0).max(65535);

const RouteSchema = z.strictObject({
  ...ModifiedIndex,
  ...Metadata,
  uris: z.array(z.string()).min(1),
  hosts: z.array(z.string()).min(1).optional(),
  methods: z
    .array(
      z.enum([
        'GET',
        'POST',
        'PUT',
        'DELETE',
        'PATCH',
        'HEAD',
        'OPTIONS',
        'CONNECT',
        'TRACE',
        'PURGE',
      ]),
    )
    .optional(),
  remote_addrs: z.array(z.string()).min(1).optional(),
  vars: z.array(z.unknown()).optional(),
  filter_func: z.string().optional(),

  plugins: Plugins.optional(),
  service_id: Metadata.id,

  timeout: Timeout.optional(),
  enable_websocket: z.boolean().optional(),
  priority: z.int().optional(),
  status: Status.optional(),
});
export type Route = z.infer<typeof RouteSchema>;

const upstreamHealthCheckPassiveHealthy = z.strictObject({
  http_statuses: z
    .array(z.coerce.number().int().min(200).max(599))
    .min(1)
    .default([200, 302])
    .optional(),
  successes: z.coerce.number().int().min(1).max(254).default(2).optional(),
});
const upstreamHealthCheckPassiveUnhealthy = z.strictObject({
  http_statuses: z
    .array(z.coerce.number().int().min(200).max(599))
    .min(1)
    .default([429, 404, 500, 501, 502, 503, 504, 505])
    .optional(),
  http_failures: z.coerce.number().int().min(1).max(254).default(5).optional(),
  tcp_failures: z.coerce.number().int().min(1).max(254).default(2).optional(),
  timeouts: z.coerce.number().int().min(1).max(254).default(3).optional(),
});
const upstreamHealthCheckType = z
  .union([z.literal('http'), z.literal('https'), z.literal('tcp')])
  .default('http');
const UpstreamSchema = z.strictObject({
  ...ModifiedIndex,
  ...Metadata,
  nodes: z
    .array(
      z.strictObject({
        host: z.string(),
        port: Port,
        weight: z.int(),
        priority: z.int().optional(),
        metadata: z.record(z.string(), z.unknown()).optional(),
      }),
    )
    .optional(),
  scheme: z
    .union([
      z.literal('http'),
      z.literal('https'),
      z.literal('grpc'),
      z.literal('grpcs'),
      z.literal('tcp'),
      z.literal('udp'),
      z.literal('tls'),
      z.literal('kafka'),
    ])
    .optional(),
  type: z
    .union([
      z.literal('roundrobin'),
      z.literal('chash'),
      z.literal('least_conn'),
      z.literal('ewma'),
    ])
    .optional(),
  hash_on: z.string().optional(),
  key: z.string().optional(),

  pass_host: z
    .union([z.literal('pass'), z.literal('node'), z.literal('rewrite')])
    .optional(),
  upstream_host: z.string().optional(),
  retries: z.int().optional(),
  retry_timeout: z.coerce.number().optional(),
  timeout: Timeout.optional(),
  tls: z
    .strictObject({
      client_cert_id: z.string().optional(),
      client_cert: z.string().optional(),
      client_key: z.string().optional(),
      verify: z.boolean().optional(),
    })
    .optional(),
  keepalive_pool: z
    .strictObject({
      size: z.int(),
      idle_timeout: z.int(),
      requests: z.int(),
    })
    .optional(),

  checks: z
    .strictObject({
      active: z.strictObject({
        type: upstreamHealthCheckType.optional(),
        timeout: z.coerce.number().default(1).optional(),
        concurrency: z.coerce.number().default(10).optional(),
        host: z.string().min(1).optional(),
        port: z.coerce.number().int().min(1).max(65535).optional(),
        http_path: z.string().default('/').optional(),
        https_verify_cert: z.boolean().default(true).optional(),
        http_request_headers: z.array(z.string()).min(1).optional(),
        healthy: z
          .strictObject({
            ...upstreamHealthCheckPassiveHealthy.shape,
            interval: z.coerce.number().int().min(1).default(1),
          })
          .optional(),
        unhealthy: z
          .strictObject({
            ...upstreamHealthCheckPassiveUnhealthy.shape,
            interval: z.coerce.number().int().min(1).default(1),
          })
          .optional(),
      }),
      passive: z
        .strictObject({
          type: upstreamHealthCheckType.optional(),
          healthy: upstreamHealthCheckPassiveHealthy.optional(),
          unhealthy: upstreamHealthCheckPassiveUnhealthy.optional(),
        })
        .optional(),
    })
    .optional(),
  discovery_type: z.string().optional(),
  service_name: z.string().optional(),
  discovery_args: z.record(z.string(), z.any()).optional(),
});
export type Upstream = z.infer<typeof UpstreamSchema>;

const ServiceSchema = z.strictObject({
  ...ModifiedIndex,
  ...Metadata,
  hosts: z.array(z.string()).min(1).optional(),
  upstream_id: z.string().optional(),
  plugins: Plugins.optional(),
});
export type Service = z.infer<typeof ServiceSchema>;

const ConsumerSchema = z.strictObject({
  ...ModifiedIndex,
  username: z.string(),
  desc: Metadata.desc,
  labels: Metadata.labels,

  plugins: Plugins.optional(),
});
export type Consumer = z.infer<typeof ConsumerSchema>;

const ConsumerCredentialSchema = z.strictObject({
  ...ModifiedIndex,
  ...Metadata,
  plugins: Plugins.optional(),
});
export type ConsumerCredential = z.infer<typeof ConsumerCredentialSchema>;

const SSLSchema = z
  .strictObject({
    ...ModifiedIndex,
    id: Metadata.id,
    desc: Metadata.desc,
    labels: Metadata.labels,

    type: z.union([z.literal('server'), z.literal('client')]).optional(),
    snis: z.array(z.string()).min(1),
    cert: z.string(),
    key: z.string(),
    certs: z.array(z.string()).optional(),
    keys: z.array(z.string()).optional(),
    client: z
      .strictObject({
        ca: z.string(),
        depth: z.int(),
        skip_mtls_uri_regex: z.array(z.string()).optional(),
      })
      .optional(),
    ssl_protocols: z
      .array(z.enum(['TLSv1.1', 'TLSv1.2', 'TLSv1.3']))
      .optional(),
    status: Status.optional(),
  })
  .extend(ModifiedIndex);
export type SSL = z.infer<typeof SSLSchema>;

const StreamRouteSchema = z.strictObject({
  ...ModifiedIndex,
  ...Metadata,
  remote_addr: z.string().optional(),
  server_addr: z.string().optional(),
  server_port: Port.optional(),
  sni: z.string().optional(),
  service_id: Metadata.id,
  plugins: Plugins.optional(),
  protocol: z
    .strictObject({
      name: z.string(),
      superior_id: z.string().optional(),
      conf: z.record(z.string(), z.unknown()).optional(),
      logger: z
        .array(
          z.strictObject({
            conf: z.record(z.string(), z.unknown()),
            name: z.string().optional(),
            filter: z.array(z.unknown()).optional(),
          }),
        )
        .optional(),
    })
    .optional(),
});
export type StreamRoute = z.infer<typeof StreamRouteSchema>;

const GlobalRuleSchema = z.strictObject({
  ...ModifiedIndex,
  id: Metadata.id,
  plugins: Plugins.optional(),
});
export type GlobalRule = z.infer<typeof GlobalRuleSchema>;

const PluginMetadataSchema = z.looseObject({
  ...ModifiedIndex,
  id: Metadata.id,
  // and arbitrary kv pairs
});
export type PluginMetadata = z.infer<typeof PluginMetadataSchema>;

export type UsedResourceTypes =
  | ADCSDK.ResourceType.ROUTE
  | ADCSDK.ResourceType.SERVICE
  | ADCSDK.ResourceType.CONSUMER
  | ADCSDK.ResourceType.SSL
  | ADCSDK.ResourceType.GLOBAL_RULE
  | ADCSDK.ResourceType.PLUGIN_METADATA
  | ADCSDK.ResourceType.UPSTREAM
  | ADCSDK.ResourceType.STREAM_ROUTE;

export const APISIXStandaloneKeyMap: {
  [K in UsedResourceTypes]: K extends ADCSDK.ResourceType.PLUGIN_METADATA
    ? `${K}`
    : `${K}s`;
} = {
  [ADCSDK.ResourceType.ROUTE]: 'routes',
  [ADCSDK.ResourceType.SERVICE]: 'services',
  [ADCSDK.ResourceType.CONSUMER]: 'consumers',
  [ADCSDK.ResourceType.SSL]: 'ssls',
  [ADCSDK.ResourceType.GLOBAL_RULE]: 'global_rules',
  [ADCSDK.ResourceType.PLUGIN_METADATA]: 'plugin_metadata',
  [ADCSDK.ResourceType.UPSTREAM]: 'upstreams',
  [ADCSDK.ResourceType.STREAM_ROUTE]: 'stream_routes',
} as const;

export const APISIXStandaloneConfVersionKeyMap: {
  [K in keyof typeof APISIXStandaloneKeyMap]: `${(typeof APISIXStandaloneKeyMap)[K]}_conf_version`;
} = {
  [ADCSDK.ResourceType.ROUTE]: 'routes_conf_version',
  [ADCSDK.ResourceType.SERVICE]: 'services_conf_version',
  [ADCSDK.ResourceType.CONSUMER]: 'consumers_conf_version',
  [ADCSDK.ResourceType.SSL]: 'ssls_conf_version',
  [ADCSDK.ResourceType.GLOBAL_RULE]: 'global_rules_conf_version',
  [ADCSDK.ResourceType.PLUGIN_METADATA]: 'plugin_metadata_conf_version',
  [ADCSDK.ResourceType.UPSTREAM]: 'upstreams_conf_version',
  [ADCSDK.ResourceType.STREAM_ROUTE]: 'stream_routes_conf_version',
} as const;

type ResourceFor<T extends UsedResourceTypes> =
  T extends ADCSDK.ResourceType.ROUTE
    ? typeof RouteSchema
    : T extends ADCSDK.ResourceType.SERVICE
      ? typeof ServiceSchema
      : T extends ADCSDK.ResourceType.CONSUMER
        ? z.ZodUnion<
            readonly [typeof ConsumerSchema, typeof ConsumerCredentialSchema]
          >
        : T extends ADCSDK.ResourceType.SSL
          ? typeof SSLSchema
          : T extends ADCSDK.ResourceType.GLOBAL_RULE
            ? typeof GlobalRuleSchema
            : T extends ADCSDK.ResourceType.PLUGIN_METADATA
              ? typeof PluginMetadataSchema
              : T extends ADCSDK.ResourceType.UPSTREAM
                ? typeof UpstreamSchema
                : T extends ADCSDK.ResourceType.STREAM_ROUTE
                  ? typeof StreamRouteSchema
                  : never;

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const APISIXStandaloneSchema = z.strictObject({
  ...({
    [APISIXStandaloneKeyMap[ADCSDK.ResourceType.ROUTE]]: z
      .array(RouteSchema)
      .optional(),
    [APISIXStandaloneKeyMap[ADCSDK.ResourceType.SERVICE]]: z
      .array(ServiceSchema)
      .optional(),
    [APISIXStandaloneKeyMap[ADCSDK.ResourceType.CONSUMER]]: z
      .array(z.union([ConsumerSchema, ConsumerCredentialSchema]))
      .optional(),
    [APISIXStandaloneKeyMap[ADCSDK.ResourceType.SSL]]: z
      .array(SSLSchema)
      .optional(),
    [APISIXStandaloneKeyMap[ADCSDK.ResourceType.GLOBAL_RULE]]: z
      .array(GlobalRuleSchema)
      .optional(),
    [APISIXStandaloneKeyMap[ADCSDK.ResourceType.PLUGIN_METADATA]]: z
      .array(PluginMetadataSchema)
      .optional(),
    [APISIXStandaloneKeyMap[ADCSDK.ResourceType.UPSTREAM]]: z
      .array(UpstreamSchema.extend(ModifiedIndex))
      .optional(),
    [APISIXStandaloneKeyMap[ADCSDK.ResourceType.STREAM_ROUTE]]: z
      .array(StreamRouteSchema)
      .optional(),
  } as {
    [K in UsedResourceTypes as (typeof APISIXStandaloneKeyMap)[K]]: z.ZodOptional<
      z.ZodArray<ResourceFor<K>>
    >;
  }),
  ...(Object.fromEntries(
    Object.values(APISIXStandaloneConfVersionKeyMap).map((k) => [
      k,
      z.int().optional(),
    ]),
  ) as {
    [K in (typeof APISIXStandaloneConfVersionKeyMap)[keyof typeof APISIXStandaloneConfVersionKeyMap]]: z.ZodOptional<z.ZodInt>;
  }),
});

export type APISIXStandalone = z.infer<typeof APISIXStandaloneSchema>;

export type ServerTokenMap = Map<string, string>;

File: rust/crates/adc-backend-apisix-standalone/src/typing.rs
//! APISIX standalone's `/apisix/admin/configs` wire shape — the whole
//! declarative config document standalone mode reads/writes atomically, as
//! opposed to `adc_sdk::resources::*` (ADC's own resource model). Unlike
//! `adc-backend-apisix`'s per-collection admin API, every resource type here
//! lives as an array inside one document, each entry stamped with a
//! `modifiedIndex` version number instead of being independently versioned.
//!
//! `labels` is a plain `Record<string, string>` on every resource here
//! (never the string-or-array `Labels` union `adc_sdk::resources` and
//! `adc-backend-apisix`'s wire types use) — standalone's admin API schema
//! only ever accepts flat string values.
//!
//! `Deserialize` stays permissive (no `deny_unknown_fields`): a live
//! standalone config document carries fields this crate doesn't need to
//! model (e.g. `X-Last-Modified`/`X-Digest` metadata APISIX embeds in the
//! body itself), and an unrecognized field from a newer APISIX release
//! should be ignored, not rejected. `Serialize` omits `None` fields via
//! `skip_serializing_if` rather than sending explicit `null`s.

use std::collections::BTreeMap;

use adc_sdk::resources::{
    Expr, HttpMethod, Plugins, SslClient, SslProtocol, SslType, Timeout, UpstreamBalancer,
    UpstreamHealthCheckActiveHealthy, UpstreamHealthCheckActiveUnhealthy,
    UpstreamHealthCheckPassive, UpstreamHealthCheckType, UpstreamKeepalivePool, UpstreamNode,
    UpstreamPassHost, UpstreamScheme, UpstreamTls,
};
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

/// Bookkeeping label a service-inlined default upstream is stamped with, so
/// [`crate::transformer::to_adc`] can find "which upstreams belong to this
/// service" among the flat top-level `upstreams` array — mirrors
/// `adc-backend-apisix`'s identically-named constant (this crate does
/// depend on that one, for its `Validator`, but not for this: the two
/// crates' upstream wire shapes differ too much to share the constant's
/// usage, so it's redefined here rather than imported).
pub const ADC_UPSTREAM_SERVICE_ID_LABEL: &str = "__ADC_UPSTREAM_SERVICE_ID";

pub type StandaloneLabels = BTreeMap<String, String>;

#[derive(Debug, Clone, Default, PartialEq, Deserialize, Serialize)]
pub struct Route {
    #[serde(rename = "modifiedIndex")]
    pub modified_index: i64,
    pub id: String,
    pub name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub desc: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub labels: Option<StandaloneLabels>,

    pub uris: Vec<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub hosts: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub methods: Option<Vec<adc_sdk::resources::HttpMethod>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub remote_addrs: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub vars: Option<Expr>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub filter_func: Option<String>,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub plugins: Option<Plugins>,
    pub service_id: String,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub timeout: Option<Timeout>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub enable_websocket: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub priority: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub status: Option<i64>,
}

/// APISIX's cjson encodes an empty Lua table as a JSON object (`{}`)
/// instead of an array — a standalone config document read back after being
/// stored with an empty array can come back as `{}` where a `Vec` was
/// expected. Normalizes that (and an explicit `null`, which `#[serde(default)]`
/// alone doesn't cover — that only fills in a *missing* key, not a present
/// one holding `null`) to an empty vec instead of rejecting it.
fn array_tolerating_empty_object<T: serde::de::DeserializeOwned>(value: Value) -> Result<Vec<T>, serde_json::Error> {
    match value {
        Value::Null => Ok(Vec::new()),
        Value::Object(map) if map.is_empty() => Ok(Vec::new()),
        other => serde_json::from_value(other),
    }
}

fn deserialize_array_tolerating_empty_object<'de, D, T>(deserializer: D) -> Result<Vec<T>, D::Error>
where
    D: serde::Deserializer<'de>,
    T: serde::de::DeserializeOwned,
{
    array_tolerating_empty_object(Value::deserialize(deserializer)?).map_err(serde::de::Error::custom)
}

/// `#[serde(default)]` alone only fills in a *missing* `conf_version` key —
/// a present key holding `null` still fails to deserialize into a bare
/// `i64`. Falls back to `0` for that case too.
fn deserialize_conf_version<'de, D>(deserializer: D) -> Result<i64, D::Error>
where
    D: serde::Deserializer<'de>,
{
    Ok(Option::<i64>::deserialize(deserializer)?.unwrap_or(0))
}

fn deserialize_upstream_nodes<'de, D>(
    deserializer: D,
) -> Result<Option<Vec<UpstreamNode>>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    match Option::<Value>::deserialize(deserializer)? {
        None | Some(Value::Null) => Ok(None),
        Some(other) => array_tolerating_empty_object(other).map(Some).map_err(serde::de::Error::custom),
    }
}

/// APISIX config file's wire field is `req_headers`; ADC's is
/// `http_req_headers`. Every other field is named the same.
#[derive(Debug, Clone, Default, PartialEq, Deserialize, Serialize)]
pub struct UpstreamHealthCheckActive {
    #[serde(rename = "type", default, skip_serializing_if = "Option::is_none")]
    pub ty: Option<UpstreamHealthCheckType>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub timeout: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub concurrency: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub host: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub port: Option<u16>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub http_method: Option<HttpMethod>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub http_path: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub req_headers: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub http_req_body: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub https_verify_certificate: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub healthy: Option<UpstreamHealthCheckActiveHealthy>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub unhealthy: Option<UpstreamHealthCheckActiveUnhealthy>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize, Serialize)]
pub struct UpstreamHealthCheck {
    pub active: UpstreamHealthCheckActive,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub passive: Option<UpstreamHealthCheckPassive>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize, Serialize)]
pub struct Upstream {
    #[serde(rename = "modifiedIndex")]
    pub modified_index: i64,
    pub id: String,
    pub name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub desc: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub labels: Option<StandaloneLabels>,

    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_upstream_nodes"
    )]
    pub nodes: Option<Vec<UpstreamNode>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub scheme: Option<UpstreamScheme>,
    #[serde(rename = "type", default, skip_serializing_if = "Option::is_none")]
    pub ty: Option<UpstreamBalancer>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub hash_on: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub key: Option<String>,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub pass_host: Option<UpstreamPassHost>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub upstream_host: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub retries: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub retry_timeout: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub timeout: Option<Timeout>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tls: Option<UpstreamTls>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub keepalive_pool: Option<UpstreamKeepalivePool>,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub checks: Option<UpstreamHealthCheck>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub discovery_type: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub service_name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub discovery_args: Option<Map<String, Value>>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize, Serialize)]
pub struct Service {
    #[serde(rename = "modifiedIndex")]
    pub modified_index: i64,
    pub id: String,
    pub name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub desc: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub labels: Option<StandaloneLabels>,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub hosts: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub upstream_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub plugins: Option<Plugins>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize, Serialize)]
pub struct Consumer {
    #[serde(rename = "modifiedIndex")]
    pub modified_index: i64,
    pub username: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub desc: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub labels: Option<StandaloneLabels>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub plugins: Option<Plugins>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize, Serialize)]
pub struct ConsumerCredential {
    #[serde(rename = "modifiedIndex")]
    pub modified_index: i64,
    pub id: String,
    pub name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub desc: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub labels: Option<StandaloneLabels>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub plugins: Option<Plugins>,
}

/// `consumers[]` holds both consumers and their credentials in one flat
/// array, discriminated by which required field is present: a `Consumer`
/// always has `username`, a `ConsumerCredential` never does (it has `id` +
/// `name` instead). `#[serde(untagged)]` tries `Consumer` first; that's
/// only safe because the two shapes' required fields never overlap.
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
#[serde(untagged)]
pub enum ConsumerOrCredential {
    Consumer(Consumer),
    Credential(ConsumerCredential),
}

impl ConsumerOrCredential {
    /// The key `crate::operator::stamp_versions` matches entries by — a
    /// consumer's `username`, or a credential's `id` (already
    /// `parentId/credentials/resourceId`-shaped, see
    /// `crate::transformer::credential_to_wire`).
    pub fn identity(&self) -> &str {
        match self {
            ConsumerOrCredential::Consumer(consumer) => &consumer.username,
            ConsumerOrCredential::Credential(credential) => &credential.id,
        }
    }

    /// The `modifiedIndex` slot for whichever variant this is — lets
    /// `crate::operator::stamp_versions` treat `Vec<ConsumerOrCredential>`
    /// the same generic way it treats every other collection, instead of a
    /// bespoke case just for this one mixed-variant type.
    pub fn modified_index(&self) -> i64 {
        match self {
            ConsumerOrCredential::Consumer(consumer) => consumer.modified_index,
            ConsumerOrCredential::Credential(credential) => credential.modified_index,
        }
    }

    pub fn modified_index_mut(&mut self) -> &mut i64 {
        match self {
            ConsumerOrCredential::Consumer(consumer) => &mut consumer.modified_index,
            ConsumerOrCredential::Credential(credential) => &mut credential.modified_index,
        }
    }

    pub fn as_consumer(&self) -> Option<&Consumer> {
        match self {
            ConsumerOrCredential::Consumer(consumer) => Some(consumer),
            ConsumerOrCredential::Credential(_) => None,
        }
    }

    pub fn as_credential(&self) -> Option<&ConsumerCredential> {
        match self {
            ConsumerOrCredential::Consumer(_) => None,
            ConsumerOrCredential::Credential(credential) => Some(credential),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct Ssl {
    #[serde(rename = "modifiedIndex")]
    pub modified_index: i64,
    pub id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub desc: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub labels: Option<StandaloneLabels>,

    #[serde(rename = "type", default, skip_serializing_if = "Option::is_none")]
    pub ty: Option<SslType>,
    pub snis: Vec<String>,
    pub cert: String,
    pub key: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub certs: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub keys: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub client: Option<SslClient>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ssl_protocols: Option<Vec<SslProtocol>>,

    pub status: i64,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize, Serialize)]
pub struct GlobalRule {
    #[serde(rename = "modifiedIndex")]
    pub modified_index: i64,
    pub id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub plugins: Option<Plugins>,
}

/// A plugin's shared config: `id` + `modifiedIndex` are the only fields
/// this crate cares about; the rest of the plugin's own config keys —
/// arbitrary, shape depends on which plugin it configures — pass through
/// untouched via `extra`'s `#[serde(flatten)]`.
#[derive(Debug, Clone, Default, PartialEq, Deserialize, Serialize)]
pub struct PluginMetadata {
    #[serde(rename = "modifiedIndex")]
    pub modified_index: i64,
    pub id: String,
    #[serde(flatten)]
    pub extra: Map<String, Value>,
}

#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct StreamRouteProtocolLogger {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub filter: Option<Vec<Value>>,
    pub conf: Map<String, Value>,
}

#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct StreamRouteProtocol {
    pub name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub superior_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub conf: Option<Map<String, Value>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub logger: Option<Vec<StreamRouteProtocolLogger>>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize, Serialize)]
pub struct StreamRoute {
    #[serde(rename = "modifiedIndex")]
    pub modified_index: i64,
    pub id: String,
    pub name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub desc: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub labels: Option<StandaloneLabels>,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub remote_addr: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub server_addr: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub server_port: Option<u16>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sni: Option<String>,
    pub service_id: String,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub plugins: Option<Plugins>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub protocol: Option<StreamRouteProtocol>,
}

/// The whole `/apisix/admin/configs` document: every resource type's array,
/// plus a per-collection `${collection}_conf_version` version number bumped
/// whenever that collection changes (see `crate::operator`). Every array and
/// every `conf_version` is always present — `#[serde(default)]` tolerates an
/// older/fresh document that omits one on read, and this crate always writes
/// all 16 explicitly (`[]`/`0`, not an omitted field) rather than leaving it
/// to the reader to tell "empty" apart from "not sent".
#[derive(Debug, Clone, Default, PartialEq, Deserialize, Serialize)]
pub struct ApisixStandalone {
    #[serde(default, deserialize_with = "deserialize_array_tolerating_empty_object")]
    pub routes: Vec<Route>,
    #[serde(default, deserialize_with = "deserialize_array_tolerating_empty_object")]
    pub services: Vec<Service>,
    #[serde(default, deserialize_with = "deserialize_array_tolerating_empty_object")]
    pub consumers: Vec<ConsumerOrCredential>,
    #[serde(default, deserialize_with = "deserialize_array_tolerating_empty_object")]
    pub ssls: Vec<Ssl>,
    #[serde(default, deserialize_with = "deserialize_array_tolerating_empty_object")]
    pub global_rules: Vec<GlobalRule>,
    #[serde(default, deserialize_with = "deserialize_array_tolerating_empty_object")]
    pub plugin_metadata: Vec<PluginMetadata>,
    #[serde(default, deserialize_with = "deserialize_array_tolerating_empty_object")]
    pub upstreams: Vec<Upstream>,
    #[serde(default, deserialize_with = "deserialize_array_tolerating_empty_object")]
    pub stream_routes: Vec<StreamRoute>,

    #[serde(default, deserialize_with = "deserialize_conf_version")]
    pub routes_conf_version: i64,
    #[serde(default, deserialize_with = "deserialize_conf_version")]
    pub services_conf_version: i64,
    #[serde(default, deserialize_with = "deserialize_conf_version")]
    pub consumers_conf_version: i64,
    #[serde(default, deserialize_with = "deserialize_conf_version")]
    pub ssls_conf_version: i64,
    #[serde(default, deserialize_with = "deserialize_conf_version")]
    pub global_rules_conf_version: i64,
    #[serde(default, deserialize_with = "deserialize_conf_version")]
    pub plugin_metadata_conf_version: i64,
    #[serde(default, deserialize_with = "deserialize_conf_version")]
    pub upstreams_conf_version: i64,
    #[serde(default, deserialize_with = "deserialize_conf_version")]
    pub stream_routes_conf_version: i64,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_missing_conf_version_defaults_to_zero() {
        let wire: ApisixStandalone = serde_json::from_value(serde_json::json!({})).unwrap();
        assert_eq!(wire.routes_conf_version, 0);
    }

    #[test]
    fn an_explicit_null_conf_version_falls_back_to_zero_instead_of_failing_to_deserialize() {
        let wire: ApisixStandalone = serde_json::from_value(serde_json::json!({ "routes_conf_version": null })).unwrap();
        assert_eq!(wire.routes_conf_version, 0);
    }

    #[test]
    fn a_missing_array_field_defaults_to_empty() {
        let wire: ApisixStandalone = serde_json::from_value(serde_json::json!({})).unwrap();
        assert_eq!(wire.routes, vec![]);
    }

    #[test]
    fn an_explicit_null_array_field_falls_back_to_empty_instead_of_failing_to_deserialize() {
        let wire: ApisixStandalone = serde_json::from_value(serde_json::json!({ "routes": null })).unwrap();
        assert_eq!(wire.routes, vec![]);
    }

    #[test]
    fn a_cjson_empty_object_in_place_of_an_array_deserializes_as_empty() {
        let wire: ApisixStandalone = serde_json::from_value(serde_json::json!({ "routes": {} })).unwrap();
        assert_eq!(wire.routes, vec![]);
    }
}
